securedrop_client/crypto.py

```diff
@@ -14,64 +14,99 @@
-    out = tempfile.NamedTemporaryFile(suffix=".message")
-    err = tempfile.NamedTemporaryFile(suffix=".message-error", delete=False)
-    if is_qubes:
-        gpg_binary = "qubes-gpg-client"
-    else:
-        gpg_binary = "gpg"
-    cmd = [gpg_binary, "--decrypt", filepath]
-    res = subprocess.call(cmd, stdout=out, stderr=err)
-
-    os.unlink(filepath)  # original file
-
-    if res != 0:
-        out.close()
-        err.close()
-
-        with open(err.name) as e:
-            msg = e.read()
-            logger.error("GPG error: {}".format(msg))
-
-        os.unlink(err.name)
-        dest = ""
-    else:
-        if is_doc:
-            # Docs are gzipped, so gunzip the file
-            with gzip.open(out.name, 'rb') as infile:
-                unzipped_decrypted_data = infile.read()
-
-            # Need to split twice as filename is e.g.
-            # 1-impractical_thing-doc.gz.gpg
-            fn_no_ext, _ = os.path.splitext(
-                os.path.splitext(os.path.basename(filepath))[0])
-            dest = os.path.join(home_dir, "data", fn_no_ext)
-
-            with open(dest, 'wb') as outfile:
-                outfile.write(unzipped_decrypted_data)
-        else:
-            fn_no_ext, _ = os.path.splitext(target_filename)
-            dest = os.path.join(home_dir, "data", fn_no_ext)
-            shutil.copy(out.name, dest)
+class CryptoException(Exception):
+
+    pass
 
-        out.close()
-        err.close()
-        logger.info("Downloaded and decrypted: {}".format(dest))
 
-    return res, dest
+class GpgHelper:
+
+    def __init__(self, gpg_home: str, session, is_qubes: bool) -> None:
+        if is_qubes:  # pragma: no cover
+            gpg_binary = "qubes-gpg-client"
+        else:
+            gpg_binary = "gpg"
+        self.gpg_home = gpg_home
+        self.gpg = GPG(binary=gpg_binary, homedir=self.gpg_home)
+        self.session = session
+
+    def import_key(self, source_uuid: UUID, key_data: str) -> None:
+        local_source = self.session.query(Source).filter_by(uuid=source_uuid) \
+            .one_or_none()
+        if local_source is None:
+            raise RuntimeError('Local source not found: {}'
+                               .format(source_uuid))
+
+        res = self.gpg.import_keys(key_data)
+        if not res:
+            raise RuntimeError('Failed to import key.')
+
+        # using a set because importing a private key returns two identical
+        # fingerprints
+        fingerprints = set(res.fingerprints)
+        if len(fingerprints) != 1:
+            raise RuntimeError('Expected only one fingerprint.')
+
+        local_source.fingerprint = fingerprints.pop()
+        self.session.add(local_source)
+        self.session.commit()
+
+    def encrypt_to_source(self, source_uuid: UUID, message: str) -> str:
+        local_source = self.session.query(Source) \
+            .filter_by(uuid=source_uuid).one()
+
+        out = self.gpg.encrypt(message, local_source.fingerprint)
```

Comment:
Per irl conversation with @redshiftzero, we want to encrypt to _both_ the source key and the journalist key. We decided on an interim solution for figuring out what fingerprint to use here: upon startup, we'll try to get the fingerprint of the journalist key from `sd-gpg` (via `pretty_bad_protocol.client.list_keys()`) then add that fingerprint to the `self.gpg.encrypt()` call here.